File: corrosion.py
from fenics import *
from fenicsprecice import Adapter

mesh = Mesh('Gapmesh.xml')
x = mesh.coordinates()
x[:, :] *= 0.001 # mesh is scaled from m to mm to better fit with the existing fracture code

V = FunctionSpace(mesh, "Lagrange", 1)

def boundary(x):
    return x[0] < 0.001 + DOLFIN_EPS or x[0] > 0.0015 - DOLFIN_EPS
    
# define coupling domain as complete domain
def coupling(x):
    return True    

dim = 2
coupling_domain = AutoSubDomain(coupling)

# the FEniCS-preCICE adapter is configured via a json file. In there, we also need to specify some arbitrary (unused) read data.
# We are currently trying to get rid of this technical restriction.
precice = Adapter(adapter_config_filename="precice-adapter-config.json")

precice_dt = precice.initialize(coupling_domain, write_object=V) # here we define where the coupling should happen

    
# Define boundary condition on left and right boundary
u0 = Constant(0.0)
bc = DirichletBC(V, u0, boundary)

# Define variational problem, simply copied from FEniCS demo
u = TrialFunction(V)
v = TestFunction(V)
f = Expression("10*exp(-(pow(x[0] - 0.5, 2) + pow(x[1] - 0.5, 2)) / 0.02)", degree=2)
g = Expression("sin(5*x[0])", degree=2)
a = inner(grad(u), grad(v))*dx
L = f*v*dx + g*v*ds

while precice.is_coupling_ongoing(): # potential time loop, currently only one timestep is done

  # Compute solution
  u = Function(V)
  solve(a == L, u, bc)
  
  # Save solution in VTK format
  file = File("poisson.pvd")
  file << u
  
  # write data to preCICE and advance coupling  
  precice.write_data(u)
  precice.advance(precice_dt)

precice.finalize()

File: corrosion.cpp
#include <dolfin.h>
#include "Poisson.h"
#include "precice/SolverInterface.hpp"

using namespace dolfin;

// Sub domain for Dirichlet boundary condition
class DirichletBoundary : public SubDomain
{
  bool inside(const Array<double>& x, bool on_boundary) const
  {
    return x[0] < 0.001+DOLFIN_EPS;
  }
};

int main()
{

  auto mesh = std::make_shared<dolfin::Mesh>("Gapmesh.xml");
  
  std::vector<double> &vertices = mesh->coordinates();
  for (auto &v : vertices){
    v *= 0.001; // mesh is scaled from m to mm to better fit with the existing fracture code
  }
  
	auto V = std::make_shared<Poisson::FunctionSpace>(mesh);
	
	// Define boundary condition
	auto u0 = std::make_shared<Constant>(0.2);
	auto boundary = std::make_shared<DirichletBoundary>();
	DirichletBC bc(V, u0, boundary);

	// Define variational forms
	Poisson::BilinearForm a(V, V);
	Poisson::LinearForm L(V);
	L.f = u0;
	L.g = u0;
	
	precice::SolverInterface precice("Corrosion", "precice-config.xml", 0, 1);

  const int meshID = precice.getMeshID("Corrosion-Mesh");
  const int dataID = precice.getDataID("Gc", meshID);

  const int numberOfVertices = mesh->num_vertices();
  const int dimensions = precice.getDimensions();
  std::vector<double> writeData(numberOfVertices);
  std::vector<int> vertexIDs(numberOfVertices);
  
  precice.setMeshVertices(meshID, numberOfVertices, mesh->coordinates().data(), vertexIDs.data());

  double dt = precice.initialize();

	// solution
	Function u(V);

	// Save solution in VTK format
	File file("poisson.pvd");
	
	int i=0;

  while (precice.isCouplingOngoing()) {
		u0->operator=(0.2+i/10);
		L.f = u0;
		solve(a == L, u, bc);
		
		for (int j=0; j<numberOfVertices; j++){
		  writeData[j] = u(mesh->coordinates()[2*j], mesh->coordinates()[2*j+1]);
		}
		
    precice.writeBlockScalarData(dataID, numberOfVertices, vertexIDs.data(), writeData.data());

    dt = precice.advance(dt);
		
		file << u;
		i++;
	}
	
	precice.finalize();

	return 0;
}
